okapi-core/src/main/java/org/folio/okapi/service/impl/Storage.java

```diff
@@ -72,20 +73,31 @@ public void prepareDatabases(InitMode initModeP, Handler<ExtendedAsyncResult<Voi
       initModeP = InitMode.INIT;
     }
     final InitMode initMode = initModeP;
-    logger.info("prepareDatabases: " + initMode);
+    logger.info("prepareDatabases: {}", initMode);
 
     boolean reset = initMode != InitMode.NORMAL;
-    envStore.init(reset, res1
-      -> deploymentStore.init(reset, res2 -> {
-        if (tenantStore == null) {
-          fut.handle(new Success<>());
-        } else {
-          tenantStore.init(reset, res3
-            -> moduleStore.init(reset, fut)
-          );
-        }
-      })
-    );
+
+    Future<Void> future = Future.succeededFuture();
+    future.compose(res -> {
+      Promise promise = Promise.promise();
+      envStore.init(reset, promise.future());
+      return promise.future();
+    }).compose(res -> {
+      Promise promise = Promise.promise();
+      deploymentStore.init(reset, promise.future());
+      return promise.future();
+    }).compose(res -> {
+      Promise promise = Promise.promise();
+      tenantStore.init(reset, promise.future());
+      return promise.future();
+    }).compose(res -> {
+      if (moduleStore == null) {
+        return Future.succeededFuture();
+      }
+      Promise promise = Promise.promise();
+      moduleStore.init(reset, promise.future());
+      return promise.future();
+    }).setHandler(fut);
```

Comment:
Can you use the more idiomatic syntax explained on https://vertx.io/docs/vertx-core/java/#_sequential_composition ?

    .compose(res -> Future.future(promise -> envStore       .init(reset, (Future) promise.future())))
    .compose(res -> Future.future(promise -> deploymentStore.init(reset, (Future) promise.future())))
    .compose(res -> Future.future(promise -> tenantStore    .init(reset, (Future) promise.future())))

Reply:
Or create a util class that directly supports a sequence of compose with ExtendedAsyncResult?

Reply:
That would be nice..